designs/2019-esm-compatibilty/README.md

```diff
@@ -0,0 +1,222 @@
+
+## Motivation
+
+The new ES module standard is finally rolling out but, there already exists a massive ecosystem of packages build as CommonJS modules (incl ESLint).
+
+In the majority case, changing a package to be interpreted as an ES module by default is as simple as adding `"type": "module"` to `package.json`.
+
+In excepitonal edge cases it may require extra work to establish interoperability across package boundaries. The way ESLint uses `.js` config files is one of those edge cases.
+
+The goal of this RFC is to explore the different options for addressing this issue.
+
+## Detailed Design
+
+To understand the design outlined here requires some background info into the mechanics by which Node modules are loaded.
+
+### Jargon
+
+- Package - A NPM package (ie contains `package.json`)
+- Module - A JS file
+- CJS - A CommonJS module
+- ESM - A Standard ECMAScript module
+- Consumer - A package that uses/depends on this package
+- Dependent - Package(s) that this package needs to work
+- Boundary - The demarcation between package, consumer, dependent
+
+### A Crash Course on Package Boundaries
+
+Historically, NPM packages have come in a variety of different package formats (ie IIFE, AMD, UMD, CJS). Prior to an actual spec, NPM settled on CJS as it's de-facto module format. CJS isn't going anywhere, the addition of ESM is additive.
+
+By default, all NPM packages are CJS-based. That means all `.js` files will be read as CJS modules. To include an ES module in a CJS package, it must have the `.mjs` extension.
+
+In addition, if `package.json` contains `"type": "module"` then the package is ESM-based. Meaning, all `.js` files contained within will be treated as ESM. To include a CJS module in a ESM-based package, it must have the `.cjs` extension.
+
+This configuration option does *not* affect consumers or dependents. Their functionality depends entirely on their own configuration. Assuming all packages follow this rule, there should be no issues with interoperability between CJS/ESM.
+
+### The Scenario
+
+A user is creating a new package. They want it to be ESM for Browser <-> Node compatibility so they configure their package to be ESM-based.
+
+The user adds `eslint` as a devDependency, w/ a typical NPM script to lint the package contents.
+
+The user defines `eslint.config.js` outlining the ruleset they prefer to use within their package.
+
+### The Issue
+
+The user pacakge is ESM-based, all `.js` files within are read as ESM.
+
+ESLint is CJS-based, it loads all files within it's package boundary as CJS.
+
+The configuration file is defined as a CJS module (ie `module.exports`), but has the `.js` syntax so requiring it throws an error. ESLint, by design reaches across the package boundary to load the configuration but the user has inadvertently signaled to Node that it uses the wrong module type.
+
+### The Fix
+
+Rename `eslint.config.js -> eslint.config.cjs`
```

Comment:
Renaming is a breaking change. I guess that you intended to add `.cjs` support as the same behavior as `.js`.

Also, you need to describe the priority of each extension. What does it happen if both `.eslintrc.js` and `.eslintrc.cjs` exist at the same directory?

Reply:
Updated the names to reflect the correct naming convention.

Assuming the package is ESM-based, `.js` will evaluate to an ESModule. If it contains `module.exports` an error will be thrown. `.cjs` will always evaluate to `CommonJS` regardless of package type therefore it should receive the higher priority.